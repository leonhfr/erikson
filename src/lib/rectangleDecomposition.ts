// Packages.
import * as _ from 'lodash';
import * as decompose from 'rectangle-decomposition';
import * as turfHelpers from '@turf/helpers';

// Internal.

// Code.
export const rectangleDecomposition = (
  polygon: turfHelpers.Feature<turfHelpers.Polygon>
) => {
  const coordinates = polygon.geometry.coordinates[0].slice(0, -1);

  // Assuming turf returns a counter-clockwise polygon as the first ring
  // So passing false to the decompose clockwise flag
  const rectangles = decompose(coordinates, false);

  return rectangles;
};
