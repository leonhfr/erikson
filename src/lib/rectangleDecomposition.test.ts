// Packages.
import * as _ from 'lodash';
import * as turfHelpers from '@turf/helpers';

// Internal.
import { rectangleDecomposition } from './rectangleDecomposition';

// Code.
describe('rectangleDecomposition', () => {
  it('should return', () => {
    const pointsCollection = turfHelpers.featureCollection(
      _.map(_.uniqWith(points, _.isEqual), p => turfHelpers.point(p))
    );
    const result = rectangleDecomposition(pointsCollection);
    expect(result).toMatchSnapshot();
  });
});

const points = [[0, 44.5], [0, 42.8], [2.1, 42.8], [2.1, 44.5]].map(x =>
  x.map(Math.floor)
);
// const points = [
//   [0, 44.5],
//   [0, 42.8],
//   [2.1, 42.8],
//   [2.1, 39],
//   [10.2, 39],
//   [10.2, 40.9],
//   [4.6, 40.9],
//   [4.6, 42.8],
//   [10.2, 42.8],
//   [10.2, 44.5],
// ].map(x => x.map(Math.floor));
