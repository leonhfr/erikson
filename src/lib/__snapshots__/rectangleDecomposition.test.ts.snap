// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rectangleDecomposition should return the expected output 1`] = `
Array [
  Array [
    2,
    41.54,
    2.04,
    41.58,
  ],
  Array [
    2.04,
    41.51,
    2.29,
    41.58,
  ],
  Array [
    2.09,
    41.47,
    2.29,
    41.51,
  ],
  Array [
    2,
    41.43,
    2.09,
    41.47,
  ],
  Array [
    2.09,
    41.4,
    2.39,
    41.47,
  ],
  Array [
    2.18,
    41.36,
    2.29,
    41.4,
  ],
  Array [
    2.29,
    41.51,
    2.34,
    41.61,
  ],
]
`;
